src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -228,6 +267,186 @@ TimeSynchronizationServer & TimeSynchronizationServer::Instance()
+    char defaultNTP[kMaxDefaultNTPSize];
+    MutableCharSpan span(defaultNTP);
+    if (GetDefaultNtp(span) != CHIP_NO_ERROR)
+    {
+        emitTimeFailureEvent(kRootEndpointId);
+        return;
+    }
+    if (span.size() > kMaxDefaultNTPSize)
+    {
+        emitTimeFailureEvent(kRootEndpointId);
+        return;
+    }
+    if (GetDelegate()->UpdateTimeUsingNTPFallback(span, &mOnFallbackNTPCompletion) != CHIP_NO_ERROR)
+    {
+        emitTimeFailureEvent(kRootEndpointId);
+    }
+}
+
+#if TIME_SYNC_ENABLE_TSC_FEATURE
+void TimeSynchronizationServer::OnDeviceConnectedFn(Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
+{
+    // Connected to our trusted time source, let's read the time.
+    app::AttributePathParams readPaths[2];
+    readPaths[0] = app::AttributePathParams(kRootEndpointId, app::Clusters::TimeSynchronization::Id,
+                                            app::Clusters::TimeSynchronization::Attributes::UTCTime::Id);
+    readPaths[1] = app::AttributePathParams(kRootEndpointId, app::Clusters::TimeSynchronization::Id,
+                                            app::Clusters::TimeSynchronization::Attributes::Granularity::Id);
+
+    app::InteractionModelEngine * engine = app::InteractionModelEngine::GetInstance();
+    app::ReadPrepareParams readParams(sessionHandle);
+    readParams.mpAttributePathParamsList    = readPaths;
+    readParams.mAttributePathParamsListSize = 2;
+
+    auto attributeCache = Platform::MakeUnique<app::ClusterStateCache>(*this);
+    if (attributeCache == nullptr)
+    {
+        // This is unlikely to work if we don't have memory, but let's try
+        OnDeviceConnectionFailureFn();
+        return;
+    }
+    auto readClient = chip::Platform::MakeUnique<app::ReadClient>(engine, &exchangeMgr, attributeCache->GetBufferedCallback(),
+                                                                  app::ReadClient::InteractionType::Read);
+    if (readClient == nullptr)
+    {
+        // This is unlikely to work if we don't have memory, but let's try
+        OnDeviceConnectionFailureFn();
+        return;
+    }
+    CHIP_ERROR err = readClient->SendRequest(readParams);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Failed to read UTC time from trusted source");
+        OnDeviceConnectionFailureFn();
+        return;
+    }
+    mAttributeCache = std::move(attributeCache);
+    mReadClient     = std::move(readClient);
+}
+
+void TimeSynchronizationServer::OnDeviceConnectionFailureFn()
+{
+    // No way to read from the TrustedTimeSource, fall back to default NTP
+    AttemptToGetFallbackNTPTimeFromDelegate();
+}
+
+void TimeSynchronizationServer::OnDone(ReadClient * apReadClient)
+{
+    using namespace chip::app::Clusters::TimeSynchronization::Attributes;
+
+    Granularity::TypeInfo::Type granularity = GranularityEnum::kNoTimeGranularity;
+    mAttributeCache->Get<Granularity::TypeInfo>(kRootEndpointId, granularity);
```

Comment:
This is assuming that on failure the out param will not be modified, no?  Nothing in the API guarantees that.

Reply:
Addressed in https://github.com/project-chip/connectedhomeip/pull/29396